### PR TITLE
Lowering: update fp64 dimension checks to match Triton support

### LINKED ISSUE #35529: Severe (5-10x) performance regression in JAX/Pallas Triton kernel (JAX  0.6.2 vs 0.8.0+)
### Description

When upgrading from JAX `0.6.2` to `>=0.8.0`, I observed a reproducible **~5x to 10x performance regression** executing a Triton kernel via `pl.pallas_call`. The kernel relies heavily on a Python-level static unrolling pattern (a flat DAG) using boolean one-hot masks to perform exact tensor-contraction along arbitrary axes without using `jax.lax.slice` or `dynamic_slice` (which cause tracing/lowering issues inside Pallas).

### Benchmarks (using the script below)
Depending on the problem size.
`n_patches = 1000 * 8` (as in the code below)
*   **JAX 0.6.2**: `~0.11 ms` per execution block
*   **JAX 0.8.0**: `~0.57 ms` per execution block

`n_patches = 1000 * 64`

*   **JAX 0.6.2**: `~0.730 ms` per execution block
*   **JAX 0.8.0**: `~4.317 ms` per execution block

The regression seems related to how MLIR/Triton scheduling or register allocation handles heavily unrolled flat DAGs generated via `while/for` loops dynamically unrolling JAX array element-wise sums inside of Pallas kernels.

## Reproducible code

I have narrowed down the regression to the following `apply_1d_mask_loop`, below is a minimal reproducible example (self-contained). The problem only occurs for p=3 (tensor contraction with 8x8 matrix). Larger operators are handled via pallas dot that is faster anyway. Unfortunately, those small operators are my main goal.

```python
import time
import jax
import jax.numpy as jnp
from jax.experimental import pallas as pl
from functools import partial

# -----------------------------------------------------------------------------
# 1. The problematic exact contraction logic
# -----------------------------------------------------------------------------
def apply_1d_mask_loop(mat, x, axis):
    """
    Pallas-compatible exact contraction using Python-unrolled one-hot masks.
    Works fast in 0.6.2, but causes severe performance collapse in >= 0.8.0.
    """
    contracting_size = x.shape[axis]
    out_size = mat.shape[0]

    col_shape = [1] * x.ndim
    col_shape[axis] = out_size

    x_idx_shape = [1] * x.ndim
    x_idx_shape[axis] = contracting_size
    arange_x = jnp.arange(contracting_size, dtype=jnp.int32).reshape(x_idx_shape)
    arange_mat = jnp.arange(contracting_size, dtype=jnp.int32).reshape(1, contracting_size)

    # First iteration (k=0)
    x_mask = (arange_x == 0).astype(x.dtype)
    x_k = jnp.sum(x * x_mask, axis=axis, keepdims=True)
    mat_mask = (arange_mat == 0).astype(mat.dtype)
    mat_col = jnp.sum(mat * mat_mask, axis=1, keepdims=True).reshape(col_shape)
    result = x_k * mat_col

    # Remaining iterations dynamically unrolled in Python
    for k_idx in range(1, contracting_size):
        x_mask = (arange_x == k_idx).astype(x.dtype)
        x_k = jnp.sum(x * x_mask, axis=axis, keepdims=True)
        mat_mask = (arange_mat == k_idx).astype(mat.dtype)
        mat_col = jnp.sum(mat * mat_mask, axis=1, keepdims=True).reshape(col_shape)
        result = result + x_k * mat_col

    return result

# -----------------------------------------------------------------------------
# 2. Minimal Pallas Kernel
# -----------------------------------------------------------------------------
def pallas_patch_kernel(u_ref, f_ref, out_ref, p, block_size):
    pid = pl.program_id(0)
    start_idx = pid * block_size
    
    b_idx = pl.dslice(start_idx, block_size)
    u_patch = u_ref[b_idx, ...]
    f_patch = f_ref[b_idx, ...]
    
    P = p + 1
    # Layout transposition to 3D volumes
    u_vol = u_patch.reshape(-1, 2, 2, 2, P, P, P)
    u_vol = u_vol.transpose(0, 1, 4, 2, 5, 3, 6)
    u_patch_3d = u_vol.reshape(-1, 2 * P, 2 * P, 2 * P)
    
    f_vol = f_patch.reshape(-1, 2, 2, 2, P, P, P)
    f_vol = f_vol.transpose(0, 1, 4, 2, 5, 3, 6)
    f_patch_3d = f_vol.reshape(-1, 2 * P, 2 * P, 2 * P)

    # Static matrices
    D = jnp.ones((2 * P, 2 * P), dtype=u_patch.dtype)
    M = jnp.ones((2 * P, 2 * P), dtype=u_patch.dtype)

    # Simulated Tensor-product application
    t1 = apply_1d_mask_loop(M, apply_1d_mask_loop(M, apply_1d_mask_loop(D, u_patch_3d, 3), 2), 1)
    t2 = apply_1d_mask_loop(M, apply_1d_mask_loop(D, apply_1d_mask_loop(M, u_patch_3d, 3), 2), 1)
    t3 = apply_1d_mask_loop(D, apply_1d_mask_loop(M, apply_1d_mask_loop(M, u_patch_3d, 3), 2), 1)
    
    res_vol = (t1 + t2 + t3) - f_patch_3d

    # Reverse
    res_back = res_vol.reshape(-1, 2, P, 2, P, 2, P)
    res_back = res_back.transpose(0, 1, 3, 5, 2, 4, 6)
    correction = res_back.reshape(-1, 8, P * P * P)
    
    # Store
    out_ref[b_idx, ...] = correction

@partial(jax.jit, static_argnums=(1, 2))
def run_minimal_smoother(u_global, p, block_size):
    n_patches = u_global.shape[0]
    grid_size = n_patches // block_size
    f_global = jnp.ones_like(u_global)

    kernel = lambda u, f, out: pallas_patch_kernel(u, f, out, p=p, block_size=block_size)

    return pl.pallas_call(
        kernel,
        out_shape=jax.ShapeDtypeStruct(u_global.shape, u_global.dtype),
        grid=(grid_size,),
        interpret=False,
        backend="triton"
    )(u_global, f_global)

# -----------------------------------------------------------------------------
# 3. Execution Driver
# -----------------------------------------------------------------------------
if __name__ == "__main__":
    jax.config.update("jax_enable_x64", True)
    
    p = 3
    n_patches = 1000 * 8
    block_size = 8
    dtype = jnp.float64
    
    print(f"JAX Version: {jax.__version__}")
    u_global = jnp.ones((n_patches, 8, (p+1)**3), dtype=dtype)

    print("Compiling...")
    res = run_minimal_smoother(u_global, p, block_size).block_until_ready()
    
    n_runs = 50
    start_time = time.perf_counter()
    for _ in range(n_runs):
        res = run_minimal_smoother(u_global, p, block_size)
    res.block_until_ready()
    end_time = time.perf_counter()

    avg_time_ms = (end_time - start_time) * 1000 / n_runs
    print(f"Average execution time over {n_runs} runs: {avg_time_ms:.3f} ms")
```

### System info (python version, jaxlib version, accelerator, etc.)

- **OS**: Linux
- **JAX version (fast, baseline)**: `0.6.2`
- **JAX version (slow, regressed)**: `0.8.0` / `0.9.0`
- **Hardware**: NVIDIA A100
- **Backend**: Pallas `triton`

**Note:** The baseline results  (`0.6.2`) correspond well with the ones reported in the literature (and results by my colleague on the same GPU via CUDA).

### PR DESCRIPTION
**Pull request by mwichro**
Indirectly fixes https://github.com/jax-ml/jax/issues/35529 
The problem reported there was occurring during a small tensor contraction. 
This PR updates the  Triton lowering for `dot_general` so that MMA instructions are used. The previous assertions were too conservative -- Triton supports smaller MMAs. I have also found a bug in `jax/_src/pallas/primitives.py`, which is fixed now.

https://github.com/triton-lang/triton/pull/7310

### Details 
1.  **DType Degradation**: Fixed a bug in `jax/_src/pallas/primitives.py` where `pl.dot` would naively default the output accumulator to `float32` for non-integer types, even when inputs were `float64`. This led to compilation errors in Triton as the hardware layouts for mixed-precision `f64xf64->f32` were not correctly mapped.
2.  **Hardware Layout Constraints**: Relaxed the artificial `dim >= 16` check in `jax/_src/pallas/triton/lowering.py` and replaced it with a hardware-informed check for `float64`. While `float32/float16` can handle smaller blocks via FMA fallback, `float64` on NVIDIA MMAv2 requires a minimum of `M>=16, N>=8, K>=16` per warp tile to avoid compiler segfaults.

### Changes:
- **`jax/_src/pallas/primitives.py`**: Updated `dot` to preserve `float64` as the preferred element type when inputs are `float64`.
- **`jax/_src/pallas/triton/lowering.py`**: 
    - Removed the blanket `min(shape) < 16` restriction.
    - Added specific guards for `float64` to match NVIDIA `MMAv2` layout requirements (`M>=16, N>=8, K>=16`).
    - Added detailed comments explaining the relationship between PTX instructions and JAX/Triton register counts.
- **`tests/pallas/triton_pallas_test.py`**: Added a regression test `test_dot_fp64_valid_dimensions` to verify correct `float64` behavior on supported shapes.

### Summary:
- All tests passed on A100 GPU with `JAX_ENABLE_X64=1`.
- Verified that invalid `float64` shapes now raise a descriptive `ValueError` instead of segfaulting the compiler.
- Verified that `float32` and `float16` can now use shapes smaller than 16 (e.g., 8x8x8).

### CODE REVIEW

File: tests/pallas/triton_pallas_test.py [hunk 1]
```diff
@@ -47,9 +47,19 @@ def setUp(self):
...
     elif jtu.test_device_matches(["gpu"]):
+      is_sm80_test = any(
+          getattr(self, "_testMethodName", "").startswith(prefix)
+          for prefix in (
+              "test_dot_f32_small_dimensions",
+              "test_dot_fp64_valid_dimensions",
+              "test_dot_fp64_invalid_dimensions",
+          )
+      )
+      min_compute = "8.0" if is_sm80_test else "9.0"
```
[the remark this answers is not in the document]
Comment (mwichro) [reply tests/pallas/triton_pallas_test.py:58]: @gemini-code-assist While I agree that this would be a better way, to make it properly all the tests in that file should be restructured in this way. This change would go way beyond the scope of this PR. I would rather do it in a separate PR.

**Comment by mwichro** [comment]
The PR is ready for (human) reviews. I would appreciate running a full test suite, I could only check if it works for A100.

**Review by rdyro — APPROVED** [review]
This looks good to me, let's run the tests, thanks!

**Comment by mwichro** [comment]
The limits in fp64 MMA in this PR could be lower, but there is a bug in Triton that causes `m8n8k4.row.col.f64.f64.f64.f64` to throw a segmentation fault. I will reproduce and submit the issue to Triton, maybe even try to fix it.

I don't get why ROC tests failed. This does not look like it's related to the changes in this PR.
